server/src/test/java/org/elasticsearch/indices/SystemIndexDescriptorTests.java

```diff
@@ -0,0 +1,114 @@
+        }
+
+        {
+            Exception ex = expectThrows(IllegalArgumentException.class,
+                () -> new SystemIndexDescriptor(randomAlphaOfLength(10), randomAlphaOfLength(5)));
+            assertThat(ex.getMessage(), containsString("must start with the character [.]"));
+        }
+
+        {
+            Exception ex = expectThrows(IllegalArgumentException.class,
+                () -> new SystemIndexDescriptor(".*", randomAlphaOfLength(5)));
+            assertThat(ex.getMessage(), containsString("must not start with the character sequence [.*] to prevent conflicts"));
+        }
+        {
+            Exception ex = expectThrows(IllegalArgumentException.class,
+                () -> new SystemIndexDescriptor(".*" + randomAlphaOfLength(10), randomAlphaOfLength(5)));
+            assertThat(ex.getMessage(), containsString("must not start with the character sequence [.*] to prevent conflicts"));
+        }
+    }
+
+    public void testOverlappingPatterns() {
+        Random random = random();
+        SystemIndexDescriptor broadPattern = new SystemIndexDescriptor(".a*c*", "test");
+        SystemIndexDescriptor notOverlapping = new SystemIndexDescriptor(".bbbddd*", "test");
+        SystemIndexDescriptor overlapping1 = new SystemIndexDescriptor(".ac*", "test");
+        SystemIndexDescriptor overlapping2 = new SystemIndexDescriptor(".aaaabbbccc", "test");
+        SystemIndexDescriptor overlapping3 = new SystemIndexDescriptor(".aaabb*cccddd*", "test");
+
+        String broadPatternSource = randomAlphaOfLength(5);
+        String otherSource = randomAlphaOfLength(6);
+        Map<String, Collection<SystemIndexDescriptor>> descriptors = new HashMap<>();
+        descriptors.put(broadPatternSource, Arrays.asList(broadPattern));
+        descriptors.put(otherSource, Arrays.asList(notOverlapping, overlapping1, overlapping2, overlapping3));
+
+        IllegalStateException exception = expectThrows(IllegalStateException.class,
+            () -> SystemIndexDescriptor.checkForOverlappingPatterns(descriptors));
+        assertThat(exception.getMessage(), containsString("a system index descriptor [" + broadPattern +
+            "] from plugin [" + broadPatternSource + "] overlaps with other system index descriptors:"));
+        String fromPluginString = " from plugin [" + otherSource + "]";
+        assertThat(exception.getMessage(), containsString(overlapping1.toString() + fromPluginString));
+        assertThat(exception.getMessage(), containsString(overlapping2.toString() + fromPluginString));
+        assertThat(exception.getMessage(), containsString(overlapping3.toString() + fromPluginString));
+        assertThat(exception.getMessage(), not(containsString(notOverlapping.toString())));
+    }
+
+    /*
+     * Some overlapping patterns will not be caught by the current overlapping-pattern detection. If this changes,
+     * this may be a breaking change for the plugin API.
+     */
```

Comment:
We can get rid of this if we use limit patterns to prefix-only. I don't currently know of a use case for wildcards other than in tail position.

Reply:
This is resolved with the `Operations.intersection`-based overlapping patterns check.